File: app/src/main/java/org/example/ponycafe/StoreMainActivity.java
package org.example.ponycafe;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.Filter;
import android.widget.Filterable;
import android.widget.GridView;
import android.widget.ImageView;
import android.widget.SearchView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;
import com.squareup.picasso.Picasso;

import java.util.ArrayList;
import java.util.List;

public class StoreMainActivity extends AppCompatActivity {

    private ArrayList<MenuModal2> foodList2 = new ArrayList<MenuModal2>();
    DatabaseReference dbref;
    GridView gridView;
    MenuAdapter menuAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState){
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_store_main);

        gridView = findViewById(R.id.idGRV);
        menuAdapter = new MenuAdapter(foodList2, this);
        gridView.setAdapter(menuAdapter);

        dbref = FirebaseDatabase.getInstance().getReference("menu");
        dbref.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot snapshot) {
                for (DataSnapshot dataSnapshot : snapshot.getChildren()){
                    MenuModal2 menu = dataSnapshot.getValue(MenuModal2.class);
                    foodList2.add(menu);
                }
                menuAdapter.notifyDataSetChanged();
            }
            @Override
            public void onCancelled(@NonNull DatabaseError error) {

            }
        });
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_user, menu);
        MenuItem menuItem = menu.findItem(R.id.menu_buscar);
        SearchView searchView = (SearchView) menuItem.getActionView();
        searchView.setMaxWidth(Integer.MAX_VALUE);
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                return true;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                Log.e("TAG", "new text ==> " + newText);
                menuAdapter.getFilter().filter(newText);
                menuAdapter.notifyDataSetChanged();
                return true;
            }
        });
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()){
            case R.id.shopping_cart:
                lanzarShoppingCart();
                break;
            case R.id.user_profile:
                lanzarUserProfile();
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    private void lanzarUserProfile() {
        Intent intent = new Intent(this, UserProfile.class);
        startActivity(intent);
        finish();
    }

    private void lanzarShoppingCart() {
        Intent intent = new Intent(this, CartActivity.class);
        startActivity(intent);

    }

    public class MenuAdapter extends BaseAdapter implements Filterable {
        private ArrayList<MenuModal2> foodList2 = new ArrayList<MenuModal2>();
        private ArrayList<MenuModal2> foodListFilter = new ArrayList<MenuModal2>();
        private Context context;

        public MenuAdapter(ArrayList<MenuModal2> foodList2, Context context) {
            this.foodList2 = foodList2;
            this.foodListFilter = foodList2;
            this.context = context;
        }

        @Override
        public int getCount() {
            return foodList2.size();
        }

        @Override
        public Object getItem(int i) {
            return foodList2.get(i);
        }

        @Override
        public long getItemId(int i) {
            return i;
        }

        @Override
        public View getView(int i, View view, ViewGroup viewGroup) {

            View view1 = view;
            MenuModal2 itemsModal = foodList2.get(i);

            if (view1 == null){
                view1 = LayoutInflater.from(context).inflate(R.layout.gridview_food_item, viewGroup, false);
            }

            ImageView imageName = view1.findViewById(R.id.idIvGrid);
            TextView tvName = view1.findViewById(R.id.idTvGrid);

            String name = itemsModal.getName();
            String desc = itemsModal.getDesc();
            String img = itemsModal.getImg();
            int cost = itemsModal.getCost();

            Picasso.get().load(img).into(imageName);
            tvName.setText(name);

            view1.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    Log.e("CLICKED ITEM", "===> " + itemsModal.getName());
                    startActivity(new Intent(StoreMainActivity.this, SelectedItemActivity.class).putExtra("data", itemsModal));
                }
            });

            return view1;
        }

        @Override
        public Filter getFilter() {
            Filter filter = new Filter() {
                @Override
                protected FilterResults performFiltering(CharSequence charSequence) {
                    FilterResults filterResults = new FilterResults();
                    if (charSequence == null || charSequence.length() == 0){
                        filterResults.count = foodListFilter.size();
                        filterResults.values = foodListFilter;
                    }else {
                        String searchChr = charSequence.toString().toLowerCase();
                        ArrayList<MenuModal2> searchResult = new ArrayList<MenuModal2>();

                        for(MenuModal2 menuModal2:foodListFilter){
                            if(menuModal2.getName().toLowerCase().contains(searchChr) || menuModal2.getDesc().toLowerCase().contains(searchChr)){
                                searchResult.add(menuModal2);
                            }
                        }
                        filterResults.count = searchResult.size();
                        filterResults.values = searchResult;
                    }
                    return filterResults;
                }

                @Override
                protected void publishResults(CharSequence charSequence, FilterResults filterResults) {
                    foodList2 = (ArrayList<MenuModal2>) filterResults.values;
                    notifyDataSetChanged();
                }
            };
            return filter;
        }
    }
}

File: app/src/main/java/org/example/ponycafe/StoreMainActivity.kt
package org.example.ponycafe

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.AdapterView
import android.widget.GridView
import androidx.appcompat.app.AppCompatActivity
import android.widget.SearchView
import android.widget.Toast
import com.google.firebase.database.*
import com.google.firebase.storage.FirebaseStorage

class StoreMainActivity : AppCompatActivity() {
    lateinit var foodGRV: GridView
    lateinit var foodGRVAdapter: GridAdapter
    lateinit var foodList: ArrayList<MenuModal>
    lateinit var dbref : DatabaseReference

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_store_main)

        foodGRV = findViewById(R.id.idGRV)
        foodList = ArrayList()
        foodGRVAdapter = GridAdapter(this, foodList)
        foodGRV.adapter = foodGRVAdapter

        getMenuData()

        foodGRVAdapter.notifyDataSetChanged()
        foodGRV.onItemClickListener = AdapterView.OnItemClickListener { _, _, position, _ ->
            /*
            Toast.makeText(
                applicationContext, foodList[position].foodName + " selected",
                Toast.LENGTH_SHORT
            ).show()
            */
            val intent = Intent(this, SelectedItemActivity::class.java)
            intent.putExtra("name", foodList[position].name)
            intent.putExtra("desc", foodList[position].desc)
            intent.putExtra("cost", foodList[position].cost)
            intent.putExtra("img", foodList[position].img)
            startActivityForResult(intent,1)
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_user, menu)
        val menuItem = menu.findItem(R.id.menu_buscar)
        val searchView: SearchView = menuItem.actionView as SearchView
        searchView.maxWidth = Int.MAX_VALUE
        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String): Boolean {
                return true
            }

            override fun onQueryTextChange(newText: String): Boolean {
                Log.e("TAG", " new text ==> $newText")
                filter(newText)
                return true
            }
        })
        return true
    }

    private fun filter(text: String) {
        // creating a new array list to filter our data.
        val filteredlist: ArrayList<MenuModal> = ArrayList()
        for (item in foodList) {
            // checking if the entered string matched with any item of our recycler view.
            if (item.name?.toLowerCase()!!.contains(text.toLowerCase())) {
                // if the item is matched we are
                // adding it to our filtered list.
                filteredlist.add(item)
            }
        }
        if (filteredlist.isEmpty()) {
            // if no item is added in filtered list we are
            // displaying a toast message as no data found.
            foodGRVAdapter.filterList(filteredlist)
            Toast.makeText(this, "No Data Found..", Toast.LENGTH_SHORT).show()
        } else {
            // at last we are passing that filtered
            // list to our adapter class.
            foodGRVAdapter.filterList(filteredlist)
        }

    }
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.shopping_cart -> {
                lanzarShoppingCart()
                true
            }
            R.id.user_profile -> {
                lanzarUserProfile()
                true
            }
            else -> super.onOptionsItemSelected(item)

        }
    }

    fun lanzarUserProfile(view: View? = null) {
        val i = Intent(this, UserProfile::class.java)
        startActivity(i)
    }
    fun lanzarShoppingCart(view: View? = null) {
        val i = Intent(this, CartActivity::class.java)
        startActivity(i)
    }

    private fun getMenuData(){
        dbref = FirebaseDatabase.getInstance().getReference("menu")
        dbref.addValueEventListener(object : ValueEventListener{
            override fun onDataChange(snapshot: DataSnapshot) {
                if (snapshot.exists()){
                    for (menuSnapshot in snapshot.children){
                        val menu = menuSnapshot.getValue(MenuModal::class.java)
                        foodList.add(menu!!)
                    }
                }
                foodGRVAdapter.notifyDataSetChanged()
            }

            override fun onCancelled(error: DatabaseError) {
                TODO("Not yet implemented")
            }
        })
    }
    private fun getMenuData2(){
        dbref = FirebaseDatabase.getInstance().getReference("menu")
    }
}